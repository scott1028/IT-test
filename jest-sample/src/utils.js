export const func1 = () => {
	return 100;
}

export const func2 = () => {
	return 200;
}