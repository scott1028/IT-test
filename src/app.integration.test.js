import {
    myReducer,
    mapDispatchToProps,
    mapStoreStateToProps,
    $App
} from './app'

console.log({
    myReducer,
    mapDispatchToProps,
    mapStoreStateToProps,
    $App
})

import nock from 'nock';
import React from 'react';
import { mount } from 'enzyme'
import { createStore, bindActionCreators } from 'redux';
import { Provider, connect } from 'react-redux';

// import toJson from 'enzyme-to-json'

import colors from 'colors';


// mock something in your APP
import * as utils from './utils';
import utils2 from './utils';
console.log('[utils]', utils);
console.log('[utils2]', utils2);

// only mock utils.func1, and func2 is original.
// utils.func1.mockRestore();


beforeEach(() => {
  console.log('[beforeStarted]');
});

it('integration test sample', (done) => {
    const store = createStore(myReducer)
    const wrapper = mount(
        <Provider store={store}>
            <$App />
        </Provider>,
    )
    console.log(colors.red(wrapper.html()))
    nock('http://www.example.com').get('/a1').reply(200, { mockSuccess: true })
    wrapper.find('button').first().simulate('click')
    nock('http://www.example.com').get('/a1').reply(200, { mockSuccess: true });
    wrapper.find('button').first().simulate('click')
    nock('http://www.example.com').get('/a1').reply(200, { mockSuccess: true });
    wrapper.find('button').first().simulate('click')
    
    // trigger catch of fetch
    wrapper.find('button').first().simulate('click')

    // [first way]
    // const flushAllPromises = () => new Promise(resolve => setTimeout(resolve, 100))
    // return flushAllPromises().then(function(){
    //     console.log(colors.red(wrapper.html()))
    //     wrapper.find('button').last().simulate('click')
    //     console.log(colors.red(wrapper.html()))
    //     expect(wrapper).toMatchSnapshot()
    //     done();
    // });

    // [another way]
    var fd = setInterval(function(){
        if(wrapper.html().indexOf('Del') !== -1){
            console.log(colors.red(wrapper.html()))
            wrapper.find('button').last().simulate('click')
            console.log(colors.red(wrapper.html()))
            expect(wrapper).toMatchSnapshot()
            clearInterval(fd);
            done();
        }
        else{
            console.log(500);
        }
    }, 0);
});

it('another test', (done) => {
    console.log(utils);

    // only mock utils.func1, and func2 is original.
    jest.spyOn(utils, 'func1').mockImplementation(() => 999);

    console.log(utils.func1())

    utils.func1.mockRestore();  // this is important or you will affect next testcase
    console.log(utils.func1());

    alert(200);
    alert(400);

    // ref: https://jestjs.io/docs/en/mock-functions
    // ref: https://jestjs.io/docs/en/22.4/mock-functions  If you are using old version
    expect(alert.mock.calls.length).toBe(2);  // you invoked alert twice before
    expect(alert.mock.calls[0][0]).toBe(200);  // what argument you pass first invoking.
    expect(alert.mock.calls[1][0]).toBe(400);  // what argument you pass second invoking.

    expect(alert.mock.instances.length).toBe(2);  // equal to alert.mock.calls
    
    done();
});