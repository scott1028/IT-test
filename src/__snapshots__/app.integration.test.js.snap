// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration test sample 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(App)>
    <App
      onChange={[Function]}
      onDecrement={[Function]}
      onIncrement={[Function]}
      onRemove={[Function]}
      onTest={[Function]}
      value={
        Object {
          "todos": Array [
            1,
            2,
            3,
            4,
          ],
        }
      }
    >
      <Counter
        doOnTest={[Function]}
        onChange={[Function]}
        onDecrement={[Function]}
        onIncrement={[Function]}
        onRemove={[Function]}
        value={
          Object {
            "todos": Array [
              1,
              2,
              3,
              4,
            ],
          }
        }
      >
        <div>
          <div>
            func1: 
            99999
          </div>
          <div>
            func2: 
            200
          </div>
          <button
            onClick={[Function]}
          >
            Add
          </button>
          <button
            onClick={[Function]}
          >
            Sub
          </button>
          <button
            onClick={[Function]}
          >
            Test
          </button>
          <hr />
          <div
            key="0"
          >
            <label>
              0
            </label>
             
            <input
              onChange={[Function]}
              style={
                Object {
                  "width": 350,
                }
              }
              value={1}
            />
             
            <button
              onClick={[Function]}
            >
              Del
            </button>
          </div>
          <div
            key="1"
          >
            <label>
              1
            </label>
             
            <input
              onChange={[Function]}
              style={
                Object {
                  "width": 350,
                }
              }
              value={2}
            />
             
            <button
              onClick={[Function]}
            >
              Del
            </button>
          </div>
          <div
            key="2"
          >
            <label>
              2
            </label>
             
            <input
              onChange={[Function]}
              style={
                Object {
                  "width": 350,
                }
              }
              value={3}
            />
             
            <button
              onClick={[Function]}
            >
              Del
            </button>
          </div>
          <div
            key="3"
          >
            <label>
              3
            </label>
             
            <input
              onChange={[Function]}
              style={
                Object {
                  "width": 350,
                }
              }
              value={4}
            />
             
            <button
              onClick={[Function]}
            >
              Del
            </button>
          </div>
          <pre>
            {
  "todos": [
    1,
    2,
    3,
    4
  ]
}
          </pre>
        </div>
      </Counter>
    </App>
  </Connect(App)>
</Provider>
`;
